### PR TITLE
Filter scenarios

### LINKED ISSUE #1075: Scenario selector - The end user should be able to filter scenarios
Here's an example: I am doing a what-if analysis on a neural network by changing the parameter alpha. When I create a new scenario, "alpha 2.00", it gets put at the top of the list instead of being ordered alphabetically. The end user should be able to filter scenarios in the scenario_selector.

core_issues_1_1

### PR DESCRIPTION
**Pull request by FredLL-Avaiga**
support Reason from is_submittable
resolves #1075
`<|{scenario}|scenario_selector|multiple|filter=day.up to date;*;day.last edit date;toto|>`

### CODE REVIEW

File: frontend/taipy-gui/src/components/Taipy/TableFilter.tsx [hunk 1]
```diff
@@ -94,18 +97,17 @@ const getFilterDesc = (columns: Record<string, ColumnDesc>, colId?: string, act?
...
-                    : val;
             return {
                 col: columns[colId].dfid,
                 action: act,
-                value: typedVal,
+                value:
+                    colType == "number"
+                        ? parseFloat(val)
+                        : colType == "boolean"
+                        ? val == "1"
```
Comment (FabienLelaquais) [inline frontend/taipy-gui/src/components/Taipy/TableFilter.tsx]: No '===' on val? String is secured?

File: taipy/gui_core/_GuiCoreLib.py [hunk 2]
```diff
@@ -12,18 +12,27 @@
 import typing as t
 from datetime import datetime
 
+from taipy.core import Cycle, DataNode, Job, Scenario, Sequence, Task
 from taipy.gui import Gui
 from taipy.gui.extension import Element, ElementLibrary, ElementProperty, PropertyType
 
 from ..version import _get_version
 from ._adapters import (
+    _GCDoNotUpdate,
```
Comment (FabienLelaquais) [inline taipy/gui_core/_GuiCoreLib.py]: Not too happy with this symbol name... compared to the others.
`_GuiCoreNoUpdate`?

File: frontend/taipy/src/ScenarioViewer.tsx [hunk 3]
```diff
@@ -104,7 +104,7 @@ interface SequencesRowProps {
     onFocus: (e: MouseEvent<HTMLElement>) => void;
     focusName: string;
     setFocusName: (name: string) => void;
-    submittable: boolean;
+    submittable: boolean | string;
```
Comment (FabienLelaquais) [inline frontend/taipy/src/ScenarioViewer.tsx]: What would be the use of the Boolean case, beside a value like null or ""?

File: frontend/taipy/src/utils.ts [hunk 4]
```diff
@@ -29,7 +29,7 @@ export type ScenarioFull = [
     string[],   // authorized_tags
     boolean,    // deletable
     boolean,    // promotable
-    boolean,    // submittable
+    boolean | string,    // submittable
```
Comment (FabienLelaquais) [inline frontend/taipy/src/utils.ts]: What would be the use of the Boolean case, beside a value like null or ""?

File: taipy/gui_core/_GuiCoreLib.py [hunk 5]
```diff
@@ -12,18 +12,27 @@
...
     _GuiCoreDatanodeAdapter,
     _GuiCoreScenarioAdapter,
     _GuiCoreScenarioDagAdapter,
     _GuiCoreScenarioProperties,
 )
 from ._context import _GuiCoreContext
 
+Scenario.__bases__ += (_GCDoNotUpdate,)
```
Comment (FabienLelaquais) [inline taipy/gui_core/_GuiCoreLib.py]: 🙂

File: taipy/gui_core/_adapters.py [hunk 6]
```diff
@@ -22,21 +22,19 @@
...
     is_readable,
     is_submittable,
 )
 from taipy.core import get as core_get
+from taipy.core.config import Config
 from taipy.core.data._tabular_datanode_mixin import _TabularDataNodeMixin
 from taipy.gui._warnings import _warn
 from taipy.gui.gui import _DoNotUpdate
-from taipy.gui.utils import _TaipyBase
+from taipy.gui.utils import _is_boolean, _is_boolean_true, _TaipyBase
```
Comment (FabienLelaquais) [inline taipy/gui_core/_adapters.py]: `_is_boolean_true` sounds weird. Compared to `_is_true` for example.

File: taipy/gui_core/_adapters.py [hunk 7]
```diff
@@ -260,37 +242,94 @@ def _attr_type(attr: str):
...
+        "config id": "string",
+        "label": "string",
+        "creation date": "date",
+        "cycle label": "string",
+        "cycle start": "date",
+        "cycle end": "date",
+        "primary": "boolean",
+        "tags": "string",
+    }
+    __SC_NAMES = {
```
Comment (FabienLelaquais) [inline taipy/gui_core/_adapters.py]: I guess you meant `__SC_LABELS`
And shouldn't these be capitalized?

File: taipy/gui_core/_context.py [hunk 8]
```diff
@@ -515,11 +527,12 @@ def submit_entity(self, state: State, id: str, payload: t.Dict[str, str]):
             if sequence := data.get("sequence"):
                 entity = entity.sequences.get(sequence)
 
-            if not is_submittable(entity):
+            if not (reason := is_submittable(entity)):
                 _GuiCoreContext.__assign_var(
                     state,
                     error_var,
-                    f"{'Sequence' if sequence else 'Scenario'} {sequence or scenario_id} is not submittable.",
+                    f"{'Sequence' if sequence else 'Scenario'} {sequence or scenario_id} is not submittable. "
```
Comment (FabienLelaquais) [inline taipy/gui_core/_context.py]:
```suggestion
                    f"{'Sequence' if sequence else 'Scenario'} {sequence or scenario_id} is not submittable: "
```

**Review by FabienLelaquais — DISMISSED** [review]
💪amazing

File: frontend/taipy-gui/src/components/Taipy/TableFilter.tsx (hunk 1 above)
Comment (FredLL-Avaiga) [reply frontend/taipy-gui/src/components/Taipy/TableFilter.tsx]: I can't find how a val coul b == to "1" without being "1" nut you're right it should be ===

File: frontend/taipy/src/ScenarioViewer.tsx (hunk 3 above)
Comment (FredLL-Avaiga) [reply frontend/taipy/src/ScenarioViewer.tsx]: Indeed it could be a string

**Review by dinhlongviolin1 — DISMISSED** [review]
LGTM

**Review by dinhlongviolin1 — APPROVED** [review]

**Review by FabienLelaquais — APPROVED** [review]
👍

**Review by toan-quach — COMMENTED** [review]
lgtm!

**Pull request merged**